File: Day 048/main.py
"""
    Title: Group Anagrams Using Character Frequency Count

    Problem Statement:
        - Given a list of words, group all anagrams together.

    Optimized Approach:
        - Instead of sorting strings, count the frequency of each character.
        - Use a tuple of 26 elements (for each alphabet letter) as a hash key.
        - Store words in a dictionary where keys represent character frequencies.

    Time Complexity:
        - **O(n * m)** (n = number of words, m = average length of each word)

    Space Complexity:
        - **O(n * m)** (for storing words in dictionary)

    Author: Vishal Singhaniya
"""

from collections import defaultdict

def group_anagrams(arr):
    """
    Function to group words that are anagrams of each other.

    Parameters:
        arr (list of str): List of input words.

    Returns:
        list of list of str: Grouped anagrams.
    """

    ans = defaultdict(list)

    # Iterate through each word in the input array
    for s in arr:
        count = [0] * 26  # 26 letters in the English alphabet
        
        # Count occurrences of each character
        for c in s:
            count[ord(c) - ord("a")] += 1  
        
        # Store words in a dictionary based on character frequency
        ans[tuple(count)].append(s)  

    return list(ans.values())

def print_anagrams(groups):
    """
    Function to print grouped anagrams in a readable format.

    Parameters:
        groups (list of list of str): List containing grouped anagrams.
    """
    for group in groups:
        print("[", " ".join(group), "]")

if __name__ == "__main__":
    words = ["cat", "dog", "god", "tac", "act", "ten", "net"]
    grouped_anagrams = group_anagrams(words)

    print("Grouped Anagrams:")
    print_anagrams(grouped_anagrams)

File: Day 048/main.cpp
/*
    Title: Group Anagrams Using Hashing

    Problem Statement:
        - Given an array of strings, group the anagrams together.

    Optimized Approach:
        - Use an **unordered_map<string, vector<string>>**:
            - Sort each string alphabetically to create a unique key.
            - Store the original string in a vector mapped to this sorted key.
        - Extract all grouped anagrams from the map into a result vector.

    Time Complexity:
        - **O(n * m log m)** (where n = number of strings, m = average string length)
        - Sorting each string takes **O(m log m)**, and inserting into a map is **O(1)**.

    Space Complexity:
        - **O(n * m)** (For storing strings in the hash map)

    Author: Vishal Singhaniya
*/

#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

// Function to group anagrams together
vector<vector<string>> groupAnagram(vector<string>& arr) {
    unordered_map<string, vector<string>> mp;

    for (string s : arr) {
        string sorted_s = s;
        sort(sorted_s.begin(), sorted_s.end()); // Sort characters in the string
        mp[sorted_s].push_back(s); // Group anagrams under the same key
    }

    vector<vector<string>> res;
    for (auto& p : mp) {
        res.push_back(p.second);
    }
    
    return res;
}

// Function to print grouped anagrams
void printAnagram(const vector<vector<string>>& groups) {
    for (const auto& group : groups) {
        cout << "[ ";
        for (const string& word : group) {
            cout << word << " ";
        }
        cout << "]\n";
    }
}

int main() {
    vector<string> words = {"cat", "dog", "god", "tac", "act", "ten", "net"};

    vector<vector<string>> groupedAnagrams = groupAnagram(words);
    cout << "Grouped Anagrams: \n";
    printAnagram(groupedAnagrams);

    return 0;
}
